firedrake/eigensolver.py

```diff
@@ -0,0 +1,169 @@
+from firedrake.petsc import OptionsManager, flatten_parameters
+from firedrake.exceptions import ConvergenceError
+try:
+    from slepc4py import SLEPc
+except ImportError:
+    SLEPc = None
+__all__ = ["LinearEigenproblem",
+           "LinearEigensolver"]
+
+
+class LinearEigenproblem():
+    """Linear eigenvalue problem
+
+    The problem has the form::
+
+        A(u, v) = λ * M(u, v)
+
+    Parameters
+    ----------
+    A : ufl.Form
+        the bilinear form A(u, v)
+    M : ufl.Form
+        the mass form M(u, v),  defaults to u * v * dx.
+    bcs : DirichletBC or list of DirichletBC
+        the boundary conditions
+    bc_shift: float
+        the value to shift the boundary condition eigenvalues by.
+
+    Notes
+    -----
+
+    If Dirichlet boundary conditions are supplied then these will result in the
+    eigenproblem having a nullspace spanned by the basis functions with support
+    on the boundary. To facilitate solution, this is shifted by the specified
+    amount. It is the user's responsibility to ensure that the shift is not
+    close to an actual eigenvalue of the system.
+    """
+    def __init__(self, A, M=None, bcs=None, bc_shift=666.0):
+        if not SLEPc:
+            raise ImportError(
+                "Unable to import SLEPc, eigenvalue computation not possible "
+                "(try firedrake-update --slepc)"
+            )
+
+        self.A = A  # LHS
+        args = A.arguments()
+        v, u = args[0], args[1]
+        if M:
+            self.M = M
+        else:
+            from ufl import inner, dx
+            self.M = inner(u, v) * dx
+        self.output_space = u.function_space()
+        self.bcs = bcs
+        self.bc_shift = bc_shift
+
+    def dirichlet_bcs(self):
+        """Return an iterator over the Dirichlet boundary conditions."""
+        for bc in self.bcs:
+            yield from bc.dirichlet_bcs()
+
+    @utils.cached_property
+    def dm(self):  # cargo cult
+        r"""Return the dm associated with the output space."""
+        return self.output_space.dm
+
+
+class LinearEigensolver(OptionsManager):
+    r"""Solve a LinearEigenproblem.
+
+    Parameters
+    ----------
+    problem : LinearEigenproblem
+        The eigenproblem to solve.
+    n_evals : int
+        The number of eigenvalues to compute.
+    options_prefix : str
+        The options prefix to use for the eigensolver.
+    solver_parameters : dict
+        PETSc options for the eigenvalue problem.
+    ncv : int
+        Maximum dimension of the subspace to be used by the solver. See
+        `SLEPc.EPS.setDimensions`.
+    mpd : int
+        Maximum dimension allowed for the projected problem. See
+        `SLEPc.EPS.setDimensions`.
+    """
+
+    DEFAULT_EPS_PARAMETERS = {"eps_gen_non_hermitian": None,
+                              "st_pc_factor_shift_type": "NONZERO",
+                              "eps_type": "krylovschur",
+                              "eps_largest_imaginary": None,
```

Comment:
Maybe we do not need to include this item in `DEFAULT_EPS_PARAMETERS`. Otherwise, I think we would have to remove this item explicitly when user provided one of the mutually exclusive items: https://slepc.upv.es/documentation/current/docs/manualpages/EPS/EPSSetWhichEigenpairs.html.

Reply:
I mean this item: `"eps_largest_imaginary": None,`